mono/metadata/w32process.c

```diff
@@ -124,137 +124,97 @@ get_process_module_class (void)
 		system_image, "System.Diagnostics", "ProcessModule");
 }
 
-static guint32
-unicode_chars (const gunichar2 *str)
-{
-	guint32 len;
-	for (len = 0; str [len] != '\0'; ++len) {}
-	return len;
-}
-
 static void
-process_set_field_object (MonoObject *obj, const gchar *fieldname, MonoObject *data)
+process_set_field_ref (MonoObjectHandle obj, const char *fieldname, MonoObjectHandle data)
 {
-	MonoClass *klass;
-	MonoClassField *field;
-
-	LOGDEBUG (g_message ("%s: Setting field %s to object at %p", __func__, fieldname, data));
-
-	klass = mono_object_class (obj);
+	// FIXME Cache offsets. Per-class.
+	MonoClass *klass = mono_handle_class (obj);
 	g_assert (klass);
 
-	field = mono_class_get_field_from_name_full (klass, fieldname, NULL);
+	MonoClassField *field = mono_class_get_field_from_name_full (klass, fieldname, NULL);
 	g_assert (field);
 
-	mono_gc_wbarrier_generic_store_internal (((char *)obj) + field->offset, data);
+	MONO_HANDLE_SET_FIELD_REF (obj, field, data);
 }
 
 static void
-process_set_field_string (MonoObject *obj, const gchar *fieldname, const gunichar2 *val, guint32 len, MonoError *error)
+process_set_field_object (MonoObjectHandle obj, const char *fieldname, MonoObjectHandle data)
 {
-	MonoDomain *domain;
-	MonoClass *klass;
-	MonoClassField *field;
-	MonoString *string;
+	LOGDEBUG (g_message ("%s: Setting field %s to object at %p", __func__, fieldname, data));
 
-	error_init (error);
+	process_set_field_ref (obj, fieldname, data);
+}
 
+static void
+process_set_field_string (MonoObjectHandle obj, const char *fieldname, MonoStringHandle string)
+{
+	process_set_field_ref (obj, fieldname, MONO_HANDLE_CAST (MonoObject, string));
+}
+
+static void
+process_set_field_utf16 (MonoObjectHandle obj, MonoStringHandle str, const char *fieldname, const gunichar2 *val, guint32 len, MonoError *error)
+{
 	LOGDEBUG (g_message ("%s: Setting field %s to [%s]", __func__, fieldname, g_utf16_to_utf8 (val, len, NULL, NULL, NULL)));
 
-	domain = mono_object_domain (obj);
+	MonoDomain *domain = MONO_HANDLE_DOMAIN (obj);
 	g_assert (domain);
 
-	klass = mono_object_class (obj);
-	g_assert (klass);
-
-	field = mono_class_get_field_from_name_full (klass, fieldname, NULL);
-	g_assert (field);
-
-	string = mono_string_new_utf16_checked (domain, val, len, error);
+	mono_string_new_utf16_assign (str, domain, val, len, error);
```

Comment:
`MONO_HANDLE_ASSIGN (str, mono_string_new_utf16_handle (domain, val, len, error))` and wrap the whole thing in `HANDLE_FUNCTION_ENTER`/`HANDLE_FUNCTION_RETURN`.  Same in several other places.

Reply:
That allocates excess handles and continues to encourage doing so.
There should be primitives for easily being efficient instead of primitives only for inefficient.

Reply:
That also takes away `return`, which we are used to, but is nice to avoid where possible.

Reply:
It is disappointing that there is preference to do this less efficiently when it is easy to do better.
See https://github.com/mono/mono/pull/15789.